File: example/java/jimmer-sql-static/src/main/java/org/babyfish/jimmer/sql/example/bll/TreeService.java
package org.babyfish.jimmer.sql.example.bll;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.babyfish.jimmer.sql.example.dal.TreeNodeRepository;
import org.babyfish.jimmer.sql.example.model.TreeNode;
import org.babyfish.jimmer.sql.example.model.TreeNodeDraft;
import org.babyfish.jimmer.sql.example.model.dto.RecursiveTree;
import org.babyfish.jimmer.sql.example.model.dto.RecursiveTreeInput;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@Transactional
@RequestMapping("/tree")
public class TreeService {

    private static final Logger LOGGER = LoggerFactory.getLogger(TreeService.class);

    private final TreeNodeRepository treeNodeRepository;

    private final ObjectMapper objectMapper;

    public TreeService(TreeNodeRepository treeNodeRepository, ObjectMapper objectMapper) {
        this.treeNodeRepository = treeNodeRepository;
        this.objectMapper = objectMapper;
    }

    @GetMapping("/roots/recursive")
    public List<RecursiveTree> findRootTrees(@RequestParam(required = false) String rootName) {
        return treeNodeRepository.findByParentIsNullAndName(rootName, RecursiveTree.class);
    }

    @PutMapping("/root/recursive")
    public TreeNode saveTree(@RequestBody RecursiveTreeInput input) {

        TreeNode rootNode = input.toEntity();

        // Make sure `TreeNode.parent` is loaded because a part of key
        rootNode = TreeNodeDraft.$.produce(rootNode, draft -> draft.setParent((TreeNode) null));

        return treeNodeRepository.save(rootNode);
    }

    @DeleteMapping("/tree/{id}")
    public void deleteTree(@PathVariable long id) {
        treeNodeRepository.deleteById(id);
    }
}

File: example/kotlin/jimmer-sql-static-kt/src/main/kotlin/org/babyfish/jimmer/sql/example/bll/TreeService.kt
package org.babyfish.jimmer.sql.example.bll

import org.babyfish.jimmer.kt.new
import org.babyfish.jimmer.sql.example.dal.TreeNodeRepository
import org.babyfish.jimmer.sql.example.model.TreeNode
import org.babyfish.jimmer.sql.example.model.by
import org.babyfish.jimmer.sql.example.model.dto.RecursiveTree
import org.babyfish.jimmer.sql.example.model.dto.RecursiveTreeInput
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/tree")
@Transactional
class TreeService(
    private val treeNodeRepository: TreeNodeRepository
) {

    @GetMapping("/roots/recursive")
    fun findRootTrees(
        @RequestParam(required = false) rootName: String?
    ): List<RecursiveTree> =
        treeNodeRepository.findByParentIsNullAndName(rootName, RecursiveTree::class)

    @PutMapping("/root/recursive")
    fun saveTree(
        @RequestBody input: RecursiveTreeInput
    ): TreeNode {

        val rootNode = new(TreeNode::class).by(
            input.toEntity()
        ) {
            // `parent` must be loaded because it is a part of key
            parent = null
        }
        return treeNodeRepository.save(rootNode)
    }

    @DeleteMapping("/{id}")
    fun deleteTree(@PathVariable id: Long) {
        treeNodeRepository.deleteById(id)
    }
}
